docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,217 @@
 ---
 title: Fetching Data
-description: Fetching data from data sources besides REST
+description: Manage connections to databases and other data sources
 ---
 
-<!--  TODO(AS4) this whole article! -->
+> Looking to fetch data from a REST API? Check out [Fetching from REST](./fetching-rest).
+
+Apollo Server can fetch data from any source you need, such as a [REST API](./fetching-rest) or a database. Your server can use any number of different data sources:
+
+```mermaid
+flowchart LR;
+  restAPI(REST API);
+  mongo(MongoDB Database);
+  sql(SQL Database);
+  subgraph ApolloServer;
+    restDataSource(RESTDataSource);
+    mongoDBDataSource(MongoDBSource);
+    sqlDataSource(SQLDBSource);
+  end
+  restDataSource --Fetches data--> restAPI;
+  mongoDBDataSource --Fetches data--> mongo;
+  sqlDataSource --Fetches data--> sql;
+  client(ApolloClient);
+  client --Sends query--> ApolloServer;
+```
+
+Because your server can use any number of various data sources, keeping your [resolvers](./resolvers) tidy becomes _even more_ important. 
+
+For this reason, we recommend creating individual data source classes to encapsulate the logic of fetching data from a particular source, providing methods that resolvers can use to access data neatly. You can additionally customize your data source classes to help with caching, deduplication, or errors while resolving operations.
+
+## Creating data source classes
+
+Your data source class can be as straightforward or complex as you need it to be. You know what data your server needs, and you can let that be the guide for the methods your class should include. 
+
+Below is an example of a data source class that connects to a database:
+
+<MultiCodeBlock>
+
+```ts title="bookings.ts"
+export class BookingsDataSource {
+  private dbConnection;
+  private context;
+
+  constructor(options: { context: { token: string } }) {
+    this.dbConnection = this.initializeDBConnection();
+    this.context = options.context;
+  }
```

Comment:
We should use `contextValue` here.

It could be interesting to make use of the `token` in our example, but not totally necessary. It does seem extraneous here if we're not using it anywhere.

Reply:
This also doesn't have to be called `contextValue` at all, we could just say the `constructor` needs a `token` as one of its options and leave it to the `context` function to pass it in correctly.

Reply:
Taking this one step further, I would suggest that `token` be passed directly as an option rather than being nested in `contextValue`.

The data source needs a token to run. The fact that the token is also embedded inside a GraphQL execution context value is not really relevant.

Reply:
Good point! Updated to:
```ts
export class BookingsDataSource {
  private dbConnection;
  private token;

  constructor(options: { token: string }) {
    this.dbConnection = this.initializeDBConnection();
    this.token = options.contextValue.token;
  }

  async initializeDBConnection() {
    // set up our database details, instantiate our connection,
    // and return that database connection
    return dbConnection;
  }

  async getBooking(bookingId) {
    const user = await getUser(this.token);
    if (user) {
      return await this.dbConnection.Booking.findByPk(bookingId);
    } else {
      // handle invalid user
    }
  }

  //... more methods for finding and creating bookings
}
```